mailparse.c

```diff
@@ -231,65 +232,63 @@ PHP_FUNCTION(mailparse_mimemessage)
 {
 	zval *object = getThis();
 	php_mimepart *part;
-	zval *zpart;
-	char *mode;
-	int mode_len;
+	zval zpart;
+	zend_string *mode;
 	zval *source = NULL;
 
-	if (zend_parse_parameters(ZEND_NUM_ARGS() TSRMLS_CC, "sz!", &mode, &mode_len, &source) == FAILURE)
+	if (zend_parse_parameters(ZEND_NUM_ARGS() TSRMLS_CC, "Sz!", &mode, &source) == FAILURE)
 		RETURN_FALSE;
 
 	/* prepare the mime part for this object */
 	part = php_mimepart_alloc(TSRMLS_C);
-	MAKE_STD_ZVAL(zpart);
-	php_mimepart_to_zval(zpart, part);
+	php_mimepart_to_zval(&zpart, part->rsrc);
 
-	zend_hash_index_update(Z_OBJPROP_P(object), 0, &zpart, sizeof(zval *), NULL);
+	zend_hash_index_update(Z_OBJPROP_P(object), 0, &zpart);
 
 	/* now check the args */
 
-	if (strcmp(mode, "new") == 0)
+	if (strcmp(mode->val, "new") == 0)
 		RETURN_TRUE;
 
 	if (source == NULL)
 		RETURN_FALSE;
-	
-	if (strcmp(mode, "var") == 0 && Z_TYPE_P(source) == IS_STRING) {
+
+	if (strcmp(mode->val, "var") == 0 && Z_TYPE_P(source) == IS_STRING) {
 		/* source is the actual message */
 		part->source.kind = mpSTRING;
 
-		*part->source.zval = *source;
-		zval_copy_ctor(part->source.zval);
-		Z_SET_REFCOUNT_P(part->source.zval, 1);
+		part->source.zval = *source;
+		zval_copy_ctor(&part->source.zval);
+		Z_SET_REFCOUNT_P(&part->source.zval, 1);
 		convert_to_string_ex(&part->source.zval);
```

Comment:
hmm, you need understand what is doing then change it,  Z_SET_REFCOUNT_P smells bad here.

and you should always use ZVAL_COPY(RES/STR) to assign a zval to another.